src/Solvers/multigrid_solver.jl

```diff
@@ -86,55 +107,145 @@ function MultigridSolver(linear_operation!::Function,
     b_array = arch_array(arch, zeros(FT, Nx * Ny * Nz))
     x_array = arch_array(arch, zeros(FT, Nx * Ny * Nz))
 
-    return MultigridSolver(arch,
-                           template_field.grid,
-                           matrix,
-                           FT(abstol),
-                           FT(reltol),
-                           maxiter,
-                           amg_algorithm,
-                           x_array,
-                           b_array
-                           )
+    arch == CPU() && return MultigridCPUSolver(arch,
+                                               template_field.grid,
+                                               matrix,
+                                               FT(abstol),
+                                               FT(reltol),
+                                               maxiter,
+                                               x_array,
+                                               b_array,
+                                               amg_algorithm
+                                               )
+
+    arch == GPU() && begin
+        try
+            AMGX.initialize()
+            AMGX.initialize_plugins()
+        catch e
+            AMGX.finalize_plugins()
+            AMGX.finalize()        
+            AMGX.initialize()
+            AMGX.initialize_plugins()
+        end
+
+        if reltol == 0
+            @info "Multigrid solver will use abstol = ($abstol)"
+            config = AMGX.Config(Dict("monitor_residual" => 1, "max_iters" => maxiter, "store_res_history" => 1, "tolerance" => abstol))
+        else
+            @info "Multigrid solver will use reltol = ($reltol)"
+            config = AMGX.Config(Dict("monitor_residual" => 1, "max_iters" => maxiter, "store_res_history" => 1, "tolerance" => reltol, "convergence" => "RELATIVE_INI_CORE"))
+        end
+        resources = AMGX.Resources(config)
+        solver = AMGX.Solver(resources, AMGX.dDDI, config)
+        device_matrix = AMGX.AMGXMatrix(resources, AMGX.dDDI)
+        device_b = AMGX.AMGXVector(resources, AMGX.dDDI)
+        device_x = AMGX.AMGXVector(resources, AMGX.dDDI)
+        csr_matrix = CuSparseMatrixCSR(transpose(matrix))
+        
+        # # FIXME do on GPU
+        # loop! = zero_index!(Architectures.device(arch), 16, length(csr_matrix.rowPtr))
+        # event = loop!(csr_matrix.rowPtr; dependencies=device_event(arch))
```

Comment:
is there something to be done here?

Reply:
Yes, I was trying to do the `subtract_one` on a gpu which should be very straight forward but I'm getting 
```
Make sure your kernel function ends in `return`, `return nothing` or `nothing`.
If the returned value is of type `Union{}`, your Julia code probably throws an exception.
Inspect the code with `@device_code_warntype` for more details.
```